### PR TITLE
Fix for Issue 459

### PR DESCRIPTION
**Pull request by JonathanBurke**
…erring type arguments without an assignment context.  Fix for issue 459

### CODE REVIEW

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 1]
```diff
@@ -361,7 +359,12 @@ private void clampToLowerBound(InferenceResult fromArgSupertypes, List<Annotated
         InferenceResult inferredFromArgEqualities = equalitiesSolver.solveEqualities(targets, argConstraints, typeFactory);
 
         Set<TypeVariable> remainingTargets =  inferredFromArgEqualities.getRemainingTargets(targets, true);
-        InferenceResult inferredFromLubs = supertypesSolver.solveFromArguments(remainingTargets, argConstraints, typeFactory);
+        InferenceResult inferredFromLubs = supertypesSolver.solveFromSupertypes(remainingTargets, argConstraints, typeFactory);
+
+        //TODO: solveFromSubtypes and then fromSubtypes (in infer) are misnamed if we use this method in this context
+        //TODO: find a clearer name
+        InferenceResult inferredFromGlbs = subtypesSolver.solveFromSubtypes(remainingTargets, argConstraints, typeFactory);
+        inferredFromLubs.mergeSubordinate(inferredFromGlbs);
```
Comment (jthaine) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: Looks good. Please do a search for similar code patterns to make sure they also handle both cases.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 2]
```diff
@@ -113,14 +111,14 @@
     public void adaptMethodType(AnnotatedTypeFactory typeFactory, ExpressionTree invocation, AnnotatedExecutableType methodType) {
         //do nothing
     }
-
+    // TODO: THIS IS A BIG VIOLATION OF Single Responsibility and SHOULD BE FIXED, IT IS SOLELY HERE
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: You moved this comment from below getArgumentTypes to inbetween adaptMethodType and getArgumentTypes - with no spacing.
To which of the two methods should it apply?
Is the whole method a kludge or something happening within it?
Which part is the single responsible party where the method should be moved to?

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 3]
```diff
@@ -298,7 +296,7 @@ private void clampToLowerBound(InferenceResult fromArgSupertypes, List<Annotated
             InferredValue inferred = fromArgSupertypes.get(targetDecl.getUnderlyingType());
             if (inferred != null && inferred instanceof InferredType) {
                 final AnnotatedTypeMirror lowerBoundAsArgument =
-                     AnnotatedTypes.asSuper(types, typeFactory, targetDecl.getLowerBound(), ((InferredType) inferred).type);
+                        AnnotatedTypes.asSuper(types, typeFactory, targetDecl.getLowerBound(), ((InferredType) inferred).type);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: Unnecessary change in spacing?

File: checker/tests/nullness/generics/Issue459.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+   {}
+
+   interface Iface<K, V>
+   {
+      public <K1 extends K, V1 extends V> Generic<K,V> foo(Generic<? super K1, ? super V1> arg);
+      
+      public <K2 extends K, V2 extends V> Generic<K,V> foo2(Generic<? super K2, ? super V2> arg, K2 strArg);
+   }
+
+   void f(Iface<Object, Object> arg, @NonNull String nnString)
```
Comment (wmdietl) [inline checker/tests/nullness/generics/Issue459.java]: The NonNull is the default for nnString.
(You could also adapt the formatting to our usual coding style and remove the unnecessary visibility modifiers.)

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 5]
```diff
@@ -361,7 +359,12 @@ private void clampToLowerBound(InferenceResult fromArgSupertypes, List<Annotated
         InferenceResult inferredFromArgEqualities = equalitiesSolver.solveEqualities(targets, argConstraints, typeFactory);
 
         Set<TypeVariable> remainingTargets =  inferredFromArgEqualities.getRemainingTargets(targets, true);
-        InferenceResult inferredFromLubs = supertypesSolver.solveFromArguments(remainingTargets, argConstraints, typeFactory);
+        InferenceResult inferredFromLubs = supertypesSolver.solveFromSupertypes(remainingTargets, argConstraints, typeFactory);
+
+        //TODO: solveFromSubtypes and then fromSubtypes (in infer) are misnamed if we use this method in this context
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: Should this be Subtypes twice? I don't see a "fromSubtypes" method or variable.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 6]
```diff
@@ -212,14 +210,14 @@ protected AnnotatedTypeMirror getAssignedTo(ExpressionTree expression, Annotated
         Pair<InferenceResult, InferenceResult> argInference = inferFromArguments(typeFactory, afArgumentConstraints, targets);
 
         final InferenceResult fromArgEqualities = argInference.first;  //result 2.a
-        final InferenceResult fromArgSupertypes = argInference.second; //result 2.b
+        final InferenceResult fromArgSubandSupers = argInference.second; //result 2.b
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: This looks like only the variable name was changed. Why is it now SubAndSuper types?
Should "clampToLowerBound" or one of the other methods called on fromArgSubandSupers be renamed?

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java (hunk 2 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: It really applies to adaptMethodType and getArgumentTypes.   TypeArgumentInference should only have the one method infer.  Unfortunately, the structure of methodFromUse/findTypeArguments etc... does not make it easy to adapt the arguments.  The single responsibility InferTypeArguments should have is to infer type arguments, it should NOT be adapting the arguments before this.  If I recall correctly, getArgumentTypes should just be replaced with it's method body and removed.  I'd have to take a closer look at SPARTA before actually removing this.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java (hunk 6 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: It is SubAndSuperTypes because those are the constraints used to create that result (rather than just Supers).  ClampToLowerBound remains unchanged.  That is, it still forces the inferred type to be at lowest the lower bound if it were inferred to be below the lower bound.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java (hunk 5 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: I have clarified the names and removed the comments.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java (hunk 3 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: This follows the Java convention.  I am not a particular fan of this convention but the previous lines also follow it.  So this will make it consistent.

**Comment by JonathanBurke** [comment]
I have responded to all of the comments in this code review.  The ones that we are still debating are stylistic.  I am going to merge this request.  Please feel free to keep commenting or send me an email if you disagree with any comment or decision I made regarding these changes.
